### PR TITLE
support single ctest

### PR DESCRIPTION
**Pull request by Alex-Lian**
use a file to specify the ctests we want to run( write the name of ctests line by line in this file)
if the file is empty, then it will automatically run all the ctests.

### CODE REVIEW

File: core/run_ctest/README.md [hunk 1]
```diff
@@ -24,7 +25,7 @@ p_input = {
 ```
 "maven_args": ["-DfailIfNoTests=false"]
 ```
-
+*For "ctest_selected", please specify the ctests you want to run in the input file, if you want to run all the ctests, just remain the file empty.
```
Comment (tianyin) [inline core/run_ctest/README.md:28]: I don't really like this design. Can we simply pass the test name as a parameter through the command line? It's kinda awkward to open a file and put the name there and then call the command (too many unneceesary keystroke)..

**Comment by tianyin** [comment]
See my commands.

File: core/run_ctest/README.md (hunk 1 above)
Comment (Alex-Lian) [reply core/run_ctest/README.md:28]: Ok

**Comment by tianyin** [comment]
@Alex-Lian Oh I realize that we will need to change the configuration parameter when running a ctest (is it the case)? If so, do we need to also support the CLI passing? Or, is it done separately?

**Comment by Alex-Lian** [comment]
> @Alex-Lian Oh I realize that we will need to change the configuration parameter when running a ctest (is it the case)? If so, do we need to also support the CLI passing? Or, is it done separately?

These two parts are separate. So my current design is using the command `./run_ctest.sh ctestname` to run a single ctest.

**Comment by tianyin** [comment]
I see. @sampalomad Is it possible for you to give a code review?

**Comment by sampalomad** [comment]
@tianyin can I create a `dev` branch and merge there?

**Comment by tianyin** [comment]
@sampalomad yes!

File: core/run_ctest/program_input.py [hunk 2]
```diff
@@ -22,4 +22,4 @@
 assert p_input["project"] \
     and p_input["mapping_path"] \
     and p_input["conf_file_dir"], ">>>>[ctest_core] please specify input"
-assert p_input["run_mode"] == "run_ctest", ">>>>[ctest_core] please specify run_mode"
+assert p_input["run_mode"] == "run_ctest" or "run_single_ctest", ">>>>[ctest_core] please specify run_mode"
```
Comment (sampalomad) [inline core/run_ctest/program_input.py:25]: @Alex-Lian I think this assertion always passes

File: core/run_ctest/main.py [hunk 3]
```diff
@@ -23,16 +23,27 @@ def main():
             print(">>>>[ctest_core] input conf file: {}".format(conf_file_path))
             test_input = extract_conf_diff(conf_file_path)
             test_conf_file(conf_file_path, test_input)
+    elif run_mode == "run_single_ctest":
+        for conf_file_path in sorted(glob.glob(os.path.join(p_input["conf_file_dir"], "*"))):
+            print(">>>>[ctest_core] input conf file: {}".format(conf_file_path))
+            test_input = extract_conf_diff(conf_file_path)
+            test_conf_file(conf_file_path, test_input, argv[1])
```
Comment (sampalomad) [inline core/run_ctest/main.py:30]: @Alex-Lian  do you want to run a single ctest on all configuration files in `conf_file_dir`, or a single ctest on any (parameter, value) you specify?

File: core/run_ctest/main.py [hunk 4]
```diff
@@ -58,4 +69,4 @@ def test_conf_file(conf_file_path, test_input):
 
 
 if __name__ == '__main__':
-    main()
+    main(sys.argv)
```
Comment (sampalomad) [inline core/run_ctest/main.py:72]: @Alex-Lian can you move arg parsing to `program_input.py` make that single ctest input an optional argument? that would be great

**Comment by tianyin** [comment]
Thank you, @sampalomad ! 

@Alex-Lian please address the review.

File: core/run_ctest/program_input.py (hunk 2 above)
Comment (Alex-Lian) [reply core/run_ctest/program_input.py]: Fixed

File: core/run_ctest/main.py (hunk 3 above)
Comment (Alex-Lian) [reply core/run_ctest/main.py:30]: The former one, in that case, we can test different groups of parameters at once. BTW, in most cases, we can remain only one file in the folder which records the parameter changes.

**Comment by tianyin** [comment]
@Alex-Lian 

>  in most cases, we can remain only one file in the folder which records the parameter changes.

I'm not a big fan of editing a file for passing information. It's simply awkward IMHO.

If we are talking about commands, what I hope is that the commands can support all what we need.

My expected usage of the commands are as following. 

1. Check all the parameters in the P set of a given ctest, say example_ctest_1
```
$ ctest show_pset example_ctest_1 
> p1
> p2
> p3
...
> p10
```
This is a TODO in https://github.com/xlab-uiuc/validator-synthesizer/issues/10 (I suggested @thakkarparth007 to take a look https://github.com/xlab-uiuc/validator-synthesizer/issues/10#issuecomment-867896525) but it's fine if you want to take it as you are more familiar with the code now.

2. run the ctest with specific parameter values
```
$ctest runtest example_ctest_1 p1=10 p2=5
```
* Basically, I really don't want to go to another file to put "p1=10 p2=5" there, but I can pass them directly to the command. It saves a lot of time and context switches.
* For p3--p10, they will just use the default value or whatever values in the test code.

Could you implement this?

**Comment by Alex-Lian** [comment]
> I'm not a big fan of editing a file for passing information. It's simply awkward IMHO.

To be honest, I also feel uncomfortable using files to pass the value....



> it's fine if you want to take it as you are more familiar with the code now

Yes, I almost finish this part.



> Could you implement this?

Sure, actually this is also my design at first, but I do not use this to avoid change more part of the original ctest.

**Comment by tianyin** [comment]
> Sure (actually this is also my design at first, but I do not use this to avoid change more part of the original ctest)

Feel free to change anything @Alex-Lian ! That's the beauty of software -- you are able to change anything in your ways :)

**Comment by Alex-Lian** [comment]
The following part is the demo.
> 1. Check all the parameters in the P set of a given ctest

image

> 2. Run the ctest with specific parameter values

image
image
image

**Comment by tianyin** [comment]
Thanks for the work, @Alex-Lian !

@thakkarparth007 Could you review the diff (which is a good exercise to get hooked on the code)?

This implements https://github.com/xlab-uiuc/validator-synthesizer/issues/10#issuecomment-867896525

File: core/run_ctest/show_pset.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import sys, json
```
Comment (thakkarparth007) [inline core/run_ctest/show_pset.py:1]: Very minor: maybe add #!/usr/env/python at the start so you can directly run the file like `show_ptest.py foobar` instead of `python3 show_ptest.py foobar`

Comment (Alex-Lian) [reply core/run_ctest/show_pset.py:1]: Thanks! But I'm worried that python will be installed in different directories for different computers.

Comment (tianyin) [reply core/run_ctest/show_pset.py:1]: `#!/usr/env/python3` (or perhaps `/usr/bin/env/python3`) should work on any reasonably system. And I believe even if the shebang does not work, you don't lose anything coz you can still call `python3` explicitly.

Comment (Alex-Lian) [reply core/run_ctest/show_pset.py:2]: I see the new version has been pushed.

**Comment by thakkarparth007** [comment]
I'll just run the code on my own setup to make sure it works as expected and then give it an "approve". Code as such looks fine.

File: core/run_ctest/run_single_ctest.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        for i in range(2, len(argv)):
+            param, value = argv[i].split('=')
+            param_value_dict[param] = value
+        test_input = extract_conf_diff_from_pair(param_value_dict)
+        test_conf_file(test_input, ctestname)
+        print(">>>>[ctest_core] total time: {} seconds".format(time.time() - s))
+
+def test_conf_file(test_input, ctestname):
+    params = test_input.keys()
+    associated_test_map = dict([[i, list((ctestname, ))] for i in params if ctestname in mapping[i]])
```
Comment (thakkarparth007) [inline core/run_ctest/run_single_ctest.py:31]: Minor: RHS could be simplified using dictionary comprehension. Took me a little while to understand what's going on

`{p: [ctestname] for p in params if ctestname in mapping[p]}`

File: core/run_ctest/run_single_ctest.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    associated_tests = set((ctestname, ))
```
Comment (thakkarparth007) [inline core/run_ctest/run_single_ctest.py:32]: Minor: Unused variable

File: core/run_ctest/run_single_ctest.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+run_mode = p_input["run_mode"]
+project = p_input["project"]
+mapping = parse_mapping(p_input["mapping_path"])
+from run_test import run_test_batch
+
+def main(argv):
+    print(">>>>[ctest_core] running project {}".format(project))
+    s = time.time()
+    ctestname = argv[1]
+    if run_mode == "run_ctest":
```
Comment (thakkarparth007) [inline core/run_ctest/run_single_ctest.py:17]: Minor: Not sure if this is needed, because we're having a special python file to run a single test. Adding this will require us to modify run_mode in program_input each time we switch between batch mode and single test mode, which can be avoided.

**Review by thakkarparth007 — APPROVED** [review]
LGTM. Added a few minor comments, but everything's working as expected. I've tested locally.
